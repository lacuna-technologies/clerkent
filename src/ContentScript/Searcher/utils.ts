export const createInput = (key: string, value: string) => {
  const newInput = document.createElement(`input`)
  newInput.setAttribute(`type`,`hidden`)
  newInput.setAttribute(`name`,key)
  newInput.setAttribute(`value`, value)
  return newInput
}

export const postFormData = (url: string, data: Record<string, string | number>[]) => {
  const form = document.createElement(`form`)
  form.setAttribute(`method`,`post`)
  form.setAttribute(`action`, url)

  for (const attribute of data) {
    const [key] = Object.keys(attribute)
    const [value] = Object.values(attribute)
    const newInput = createInput(key, `${value}`)
    form.append(newInput)
  }

  document.querySelectorAll(`body`)[0].append(form)
  form.submit()
}