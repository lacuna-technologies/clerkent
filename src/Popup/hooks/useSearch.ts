import { useCallback, useMemo, useRef, useState } from "preact/hooks"

const useSearch = () => {
  const [isSearching, setIsSearchingRaw] = useState(false)
  const [newSearchResults, setNewSearchResults] = useState([] as Law.Case[])
  const [searchResults, setSearchResults] = useState([] as Law.Case[])
  const [initialTimerDone, setInitialTimerDone] = useState(false)
  const timer = useRef(null)

  const updateResults = useCallback(() => {
    if(newSearchResults.length > 0){
      setSearchResults(newSearchResults)
      setNewSearchResults([])
    }
  }, [newSearchResults])

  const setIsSearching = useCallback((value: boolean) => {
    setIsSearchingRaw(value)

    setInitialTimerDone(false)
    if(timer?.current){
      clearTimeout(timer.current)
    }

    if(value === true){
      timer.current = setTimeout(() => {
        setInitialTimerDone(true)
        updateResults()
      }, 1000)
    }
  }, [updateResults])

  const onReceiveSearchResults = useCallback((data: { done: boolean, results: Law.Case[] }) => {
    const { done, results } = data

    if(done){
      setIsSearchingRaw(false)
      setInitialTimerDone(true)
    }

    // avoid sudden shifts in the layout
    if(searchResults.length === 0 && !initialTimerDone){
      setSearchResults(results)
      setNewSearchResults([])
    } else {
      setNewSearchResults(results)
    }
  }, [searchResults.length, setIsSearchingRaw, initialTimerDone, setInitialTimerDone])

  const updatePending = useMemo(() => (newSearchResults.length > 0), [newSearchResults])

  const resetSearchResults = useCallback(() => {
    setSearchResults([])
    setNewSearchResults([])
  }, [])

  return {
    isSearching,
    onReceiveSearchResults,
    resetSearchResults,
    searchResults,
    setIsSearching,
    updatePending,
    updateResults,
  }
}

export default useSearch