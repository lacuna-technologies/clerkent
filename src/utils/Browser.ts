import { browser } from "webextension-polyfill-ts"
import Memoize from 'memoizee'

const getExtensionLink = Memoize(
  (string = ``) => browser.runtime.getURL(string),
)
const isFirefox = () => getExtensionLink().startsWith(`moz-extension://`)
const isChrome = () => getExtensionLink().startsWith(`chrome-extension://`)

const Browser = {
  getExtensionLink,
  isChrome,
  isFirefox,
}

export default Browser
