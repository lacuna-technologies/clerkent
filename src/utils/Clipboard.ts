import { browser } from 'webextension-polyfill-ts'

const readText = async () => {
  const permissionsRequest = await browser.permissions.request({
    permissions: [`clipboardRead`],
  })
  if(permissionsRequest){
    return navigator.clipboard.readText()
  }
  return ``
}

const isURI = (string: string) => (new RegExp(/.*(:\/\/|about:|chrome:).*/, `gi`)).test(string)
const hasTooLongWord = (string: string) => {
  const matches = string.match(/\w+/g)
  if(matches === null){
    return false
  }
  return matches.sort(
    (a, b) => b.length - a.length,
  )[0].length > 30
}
const isTooLong = (string: string) => string.length > 300
const hasInvalidCharacters = (string: string) => (new RegExp(/.*[<>\\{|}].*/, `gi`)).test(string)

const getPopupSearchText = async (): Promise<string> => {
  const clipboardText = await readText()
  const isValid = (
    clipboardText && clipboardText.length > 0 &&
    !isTooLong(clipboardText) && !isURI(clipboardText) &&
    !hasTooLongWord(clipboardText) && !hasInvalidCharacters(clipboardText)
  )
  if(isValid){
    return clipboardText
  }
  return ``
}

const Clipboard = {
  getPopupSearchText,
  readText,
}

export default Clipboard