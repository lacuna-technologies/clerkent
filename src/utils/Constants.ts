import type Law from '../types/Law'

const JURISDICTIONS = {
  AU: {
    emoji: `ðŸ‡¦ðŸ‡º`,
    id: `AU` as Law.JursidictionCode,
    name: `Australia`,
  },
  CA: {
    emoji: `ðŸ‡¨ðŸ‡¦`,
    id: `CA` as Law.JursidictionCode,
    name: `Canada`,
  },
  EU: {
    emoji: `ðŸ‡ªðŸ‡º`,
    id: `EU` as Law.JursidictionCode,
    name: `European Union`,
  },
  HK: {
    emoji: `ðŸ‡­ðŸ‡°`,
    id: `HK` as Law.JursidictionCode,
    name: `Hong Kong`,
  },
  MY: {
    emoji: `ðŸ‡²ðŸ‡¾`,
    id: `MY` as Law.JursidictionCode,
    name: `Malaysia`,
  },
  NZ: {
    emoji: `ðŸ‡³ðŸ‡¿`,
    id: `NZ` as Law.JursidictionCode,
    name: `New Zealand`,
  },
  SG: {
    emoji: `ðŸ‡¸ðŸ‡¬`,
    id: `SG` as Law.JursidictionCode,
    name: `Singapore`,
  },
  UK: {
    emoji: `ðŸ‡¬ðŸ‡§`,
    id: `UK` as Law.JursidictionCode,
    name: `United Kingdom`,
  },
}


const SG_DATABASES = {
  lawnet: {
    icon: ``,
    name: `Lawnet`,
    url: `https://www.lawnet.sg/lawnet/web/lawnet/home`,
  },
  lexis: {
    icon: ``,
    name: `Lexread`,
    url: `https://www.lexread.lexisnexis.com/`,
  },
  sc: {
    icon: ``,
    name: `Supreme Court`,
    url: `https://www.supremecourt.gov.sg/news/supreme-court-judgments`,
  },
  slw: {
    icon: ``,
    name: `Singapore Law Watch`,
    url: `https://www.singaporelawwatch.sg/Judgments`,
  },
  sso: {
    icon: ``,
    name: `Singapore Statutes Online`,
    url: `https://sso.agc.gov.sg/`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw Asia`,
    url: `https://www.westlawasia.com/singapore/`,
  },
}

const UK_DATABASES = {
  bailii: {
    icon: ``,
    name: `BAILII`,
    url: `https://bailii.org/`,
  },
  legislation: {
    icon: ``,
    name: `legislation.gov.uk`,
    url: `https://www.legislation.gov.uk/`,
  },
  lexis: {
    icon: ``,
    name: `LexisNexis`,
    url: `http://www.lexisnexis.com/uk/legal`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw`,
    url: `https://uk.westlaw.com/Browse/Home/WestlawUK`,
  },
}

const EU_DATABASES = {
  curia: {
    icon: ``,
    name: `CURIA`,
    url: `https://curia.europa.eu/juris/recherche.jsf?language=en`,
  },
  epo: {
    icon: ``,
    name: `EPO`,
    url: `https://www.epo.org/law-practice/case-law-appeals`,
  },
}

const HK_DATABASES = {
  hkliihk: {
    icon: ``,
    name: `HKLII.hk`,
    url: `https://www.hklii.hk/`,
  },
  hkliiorg: {
    icon: ``,
    name: `HKLII.org`,
    url: `https://www.hklii.org/eng/`,
  },
}

const CA_DATABASES = {
  canlii: {
    icon: ``,
    name: `CanLII`,
    url: `https://www.canlii.org/en`,
  },
}

const AU_DATABASES = {
  austlii: {
    icon: ``,
    name: `AustLII`,
    url: `https://www8.austlii.edu.au`,
  },
}

const MISC_DATABASES = {
  commonlii: {
    icon: ``,
    name: `CommonLII`,
    url: `http://www.commonlii.org`,
  },
}

const dedupeObjects = (inputObject: Record<string, Record<string, string>>, jurisdictionCode: Law.JursidictionCode) =>
  Object.entries(inputObject)
  .reduce((accumulator, [id, object]: [string, Record<string, string>]) => ({
    ...accumulator,
    [`${jurisdictionCode}_${id}`]: object,
  }), {})

const DATABASES: Record<string, Law.Database> = {
  ...dedupeObjects(SG_DATABASES, `SG`),
  ...dedupeObjects(UK_DATABASES, `UK`),
  ...dedupeObjects(EU_DATABASES, `EU`),
  ...dedupeObjects(HK_DATABASES, `HK`),
  ...dedupeObjects(CA_DATABASES, `CA`),
  ...dedupeObjects(AU_DATABASES, `AU`),
  ...MISC_DATABASES,
}

const SG_COURTS = {
  sgca: {
    id: `SGCA`,
    name: `Court of Appeal`,
  },
  sghc: {
    id: `SGHC`,
    name: `High Court`,
  },
}

const EU_COURTS = {
  cjeu: {
    id: `CJEU`,
    name: `Court of Justice of the European Union`,
  },
  epo: {
    id: `EPO`,
    name: `European Patent Office`,
  },
}

const COURTS: Record<string, Record<string, string>> = {
  ...dedupeObjects(SG_COURTS, `SG`),
  ...dedupeObjects(EU_COURTS, `EU`),
}

const Constants = {
  COURTS,
  DATABASES,
  JURISDICTIONS,
}

export default Constants