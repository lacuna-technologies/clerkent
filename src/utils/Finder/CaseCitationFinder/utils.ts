import Logger from "utils/Logger"

export const formatAbbr = (
  { abbr, appendum = null }:
  { abbr: string, appendum?: string },
) => {
  const end = appendum || ``
  return `${abbr
    .split(``)
    .map((letter: string) =>
      /[a-z]/i.test(letter)
        ? `${letter}\\.?`
        : letter,
    ).join(``)
    }${end}`
}

export const formatAbbrs = (
  abbrArray: { abbr: string, appendum?: string }[],
) => abbrArray.map(
  ({ abbr, appendum }) => formatAbbr({ abbr, appendum}),
).join(`|`)

export const cleanVolume = (volumeString: string) => volumeString.replace(/\./g, ``).toLowerCase().trim()

export const sortCitationsByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: string[],
): string[] => {
  const lastIfNotFound = (index: number) => (
    index === -1
      ? citationsArray.length + 1
      : index
  )
  return citationsArray.sort((a, b) => {
    const indexA = lastIfNotFound(
      abbrsList.findIndex(
        currentAbbr => new RegExp(
          formatAbbr(currentAbbr),
          `i`,
        ).test(a),
      ),
    )
    const indexB = lastIfNotFound(
      abbrsList.findIndex(currentAbbr => 
        new RegExp(formatAbbr(currentAbbr), `i`).test(b),
      ),
    )
    return indexA - indexB
  })
}

export const sortCasesByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: Law.Case[],
  attribute = `citation`,
): Law.Case[] => sortCitationsByVolume(
    abbrsList,
    (citationsArray).map(c => c[attribute]),
  ).map(c => (citationsArray).find(v => v[attribute] === c))