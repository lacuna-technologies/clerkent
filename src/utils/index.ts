export { default as Constants } from './Constants'
export { default as Finder } from './Finder'
export { default as Messenger } from './Messenger'
export { default as Request } from './Request'
export { default as Storage } from './Storage'
export { default as Logger } from './Logger'
export { default as Helpers } from './Helpers'
export { default as Clipboard } from './Clipboard'