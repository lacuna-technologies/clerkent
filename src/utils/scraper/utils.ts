import Fuse from 'fuse.js'
import Helpers from '../Helpers'
import Constants from 'utils/Constants'
import Storage from 'utils/Storage'
import Finder from "../Finder"

export const findCitation = (function_ = Finder.findCaseCitation, inputText: string) => {
  const results = function_(inputText)
  if(results.length > 0){
    return results[0].citation
  }
  return null
}

const longestCommonSubstring = (stringA: string, stringB: string) => {
  if (!stringA || !stringB) {
    return {
      length: 0,
      str: ``,
    }
  }

  let longestSubstr = ``
  let currentSubstr = ``
  for (const element of stringA) {
    currentSubstr += element
    if (!stringB.toLowerCase().includes(currentSubstr.toLowerCase())) {
      if (currentSubstr.length > longestSubstr.length) {
        longestSubstr = currentSubstr
      }
      currentSubstr = ``
    }
  }
  return currentSubstr.length > longestSubstr.length ? {
    length: currentSubstr.length,
    str: currentSubstr,
  } : {
    length: longestSubstr.length,
    str: longestSubstr,
  }
}

const probablySameCase = (caseNameA: string, caseNameB: string) => {
  const {
    length: longestSubstrLength,
  } = longestCommonSubstring(
    caseNameA,
    caseNameB,
  )
  const sameCaseProbability = longestSubstrLength / Math.min(caseNameA.length, caseNameB.length)
  return sameCaseProbability >= 0.8
}

export const sortByName = (query: string, cases: Law.Case[]) => {
  const cleanQuery = Helpers.removeCommonAppends(query.toLowerCase())
  const fuse = new Fuse(cases.map(({ name }) => Helpers.removeCommonAppends(name.toLowerCase())), { ignoreLocation: true })
  return fuse
    .search(cleanQuery)
    .map(({ refIndex }) => cases[refIndex])
}

export const databaseUse = async (jurisdictionId: Law.JurisdictionCode, databaseId: string, function_: () => Promise<Law.Case[]>): Promise<Law.Case[]> => {
  const databasesStatus: typeof Constants.DEFAULT_DATABASES_STATUS = await Storage.get(`DATABASES_STATUS`) || Constants.DEFAULT_DATABASES_STATUS
  if(databasesStatus[jurisdictionId][databaseId]){
    return function_()
  } 
  return []
}