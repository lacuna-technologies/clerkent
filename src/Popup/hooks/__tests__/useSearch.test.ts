import { renderHook, act } from "@testing-library/preact"
import { Constants } from "utils"
import useSearch from "../useSearch"

const SEARCH_RESULTS: Law.Case[] = [
  {
    citation: `[2022] EWHC 160`,
    database: Constants.DATABASES.UK_bailii,
    jurisdiction: Constants.JURISDICTIONS.UK.id,
    links: [],
    name: `Stadler v Currys Group`,
    type: `case-citation`,
  },
  {
    citation: `[2022] EWHC 1379 (IPEC)`,
    database: Constants.DATABASES.UK_bailii,
    jurisdiction: Constants.JURISDICTIONS.UK.id,
    links: [],
    name: `Shazam Productions v Only Fools the Dining Experience`,
    type: `case-citation`,
  },
]

describe(`useSearch`, () => {
  it(`has correct initial state`, () => {
    const { result } = renderHook(() => useSearch())
    const { current: { isSearching, searchResult } } = result
    expect(isSearching).toBe(false)
    expect(searchResult.length).toBe(0)
  })
  it(`correctly sets search results`, () => {
    const { result } = renderHook(() => useSearch())
    const { current: { onSearchDone } } = result
    
    act(() => {
      onSearchDone(SEARCH_RESULTS)
    })

    const { current: { searchResult } } = result
    expect(searchResult).toMatchSnapshot()
  })
})