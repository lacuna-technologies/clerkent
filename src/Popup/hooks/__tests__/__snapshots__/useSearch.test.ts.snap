// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSearch correctly sets search results 1`] = `
Array [
  Object {
    "citation": "[2022] EWHC 160",
    "database": Object {
      "icon": "",
      "id": "UK_bailii",
      "name": "BAILII",
      "url": "https://bailii.org/",
    },
    "jurisdiction": "UK",
    "links": Array [],
    "name": "Stadler v Currys Group",
    "type": "case-citation",
  },
  Object {
    "citation": "[2022] EWHC 1379 (IPEC)",
    "database": Object {
      "icon": "",
      "id": "UK_bailii",
      "name": "BAILII",
      "url": "https://bailii.org/",
    },
    "jurisdiction": "UK",
    "links": Array [],
    "name": "Shazam Productions v Only Fools the Dining Experience",
    "type": "case-citation",
  },
]
`;
